compiler/z/codegen/S390Peephole.cpp

```diff
@@ -1349,29 +1349,89 @@ TR_S390PostRAPeephole::ConditionalBranchReduction(TR::InstOpCode::Mnemonic branc
 
 
 /**
- * Catch the pattern where an CR/BRC can be converted
- *    CR R1,R2
- *    BRC  Mask, Lable
+ * Catch the pattern where an CLR/BRC can be converted
+ *    CLR R1, R2
+ *    BRC Mask, Lable
  * Can be replaced with
- *    CRJ  R1,R2,Lable
+ *    CLRJ R1, R2, Lable, Mask
  */
 bool
 TR_S390PostRAPeephole::CompareAndBranchReduction()
    {
-   bool performed = false;
+   if (!TR::Compiler->target.cpu.getSupportsArch(TR::CPU::z10))
+      return false;
 
-   TR::Instruction *current = _cursor;
-   TR::Instruction *next = _cursor->getNext();
-   TR::InstOpCode::Mnemonic curOpCode = current->getOpCodeValue();
-   TR::InstOpCode::Mnemonic nextOpCode = next->getOpCodeValue();
-   if ((curOpCode == TR::InstOpCode::CR || curOpCode == TR::InstOpCode::CGR || curOpCode == TR::InstOpCode::CGFR)
-      && (nextOpCode == TR::InstOpCode::BRC || nextOpCode == TR::InstOpCode::BRCL))
-      {
-      printf("Finding CR + BRC\n");
-      printf("method=%s\n", comp()->signature());
+   bool branchTakenPerformReduction = false;
+   bool fallThroughPerformReduction = false;
+   TR::Instruction *clrInstruction = _cursor;
+   TR::Instruction *brcInstruction = _cursor->getNext();
+   TR::Instruction *clrjInstruction = NULL;
+   TR::LabelSymbol *labelSymbol = NULL;
+
+   if (clrInstruction->getOpCodeValue() == TR::InstOpCode::CLR
+       && brcInstruction->getOpCodeValue() == TR::InstOpCode::BRC)
+      {
+      /* if branch target is a snippet we'll go ahead with the
+       * reduction. Otherwise, we look at next 5 (or until TR::BBEnd
+       * depending on whichever comes first) to check if any
+       * instruction reads or sets the condition code.
+       * We'll perform the reduction if an instruction sets
+       * the condition code or, if none of the instructions read the
+       * condition code set by clrInstruction
+       */
+      labelSymbol = brcInstruction->getLabelSymbol();
+      if (labelSymbol->getSnippet())
+         {
+         branchTakenPerformReduction = true;
+         }
+      else
+         {
+            // check branch case for condition code usage
+            int branchTakenInstIndex = 0;
+            TR::Instruction* targetInstruction = labelSymbol->getInstruction();
+            while (branchTakenInstIndex < 5
+               && targetInstruction
+               && (branchTakenInstIndex > 0 && !targetInstruction->getPrev()->getOpCode().setsCC())
```

Comment:
Is this correct? We seem to be reading the previous instruction before the branch target. Also we will never enter this `while` loop since `branchTakenInstIndex > 0` is never true on the first execution. This is why it may potentially be better to use a `for` loop for the peephole window. Something like:

```
TR::Instruction* cursor = labelSymbol->getInstruction();
for (auto i = 0; i < 5 && cursor != NULL; ++i)
   {
   // CC is read before it is set
   if (cursor->getOpCode().readsCC())
      {
      break;
      }
      
   // CC is set before it is read (ordering of the if checks matter)
   if (cursor->getOpCode().setsCC() || cursor->getNode()->getOpCodeValue() == TR::BBEnd)
      {
       branchTakenPerformReduction = true;
       break;
      }

   cursor = cursor->getNext();
   }
```

Reply:
The idea there was to check at the start of each iteration if the previous node had set condition code. This check was to avoid having to use break statements. But that got tricky for the branch taken case as the previous node can't be reliably checked. I meant to take that `branchTakenInstIndex > 0` out before pushing

Initially I was under the impression that we shouldn't be adding any additional loop conditions in the `for` loop or using `break` statements. But since that's totally fine I'll replace the `while` loop

Reply:
`break` should be fair game. It may make the code simpler, as I think it does in this case.